x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+// distributeRewards is the main function for reward distribution for providers and delegators
+func (k Keeper) distributeRewards(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, epoch uint64, totalReward math.Int) error {
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, providerAddr, chainID, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot calculate provider reward and update delegator reward map", err)
+	}
+
+	providerRewardCoins := sdk.Coins{sdk.NewCoin(epochstoragetypes.TokenDenom, providerReward)}
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, providerAddr, providerRewardCoins)
+	if err != nil {
+		// panic:ok: reward transfer should never fail
+		utils.LavaFormatPanic("critical: failed to send reward to provider", err,
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "reward", Value: providerRewardCoins},
+		)
+	}
+
+	delegations, err := k.dualStakingKeeper.GetProviderDelegators(ctx, providerAddr.String(), epoch)
+	if err != nil {
+		return utils.LavaFormatError("cannot get provider's delegators", err)
+	}
+
+	err = k.updateDelegatorsReward(ctx, providerAddr, delegations, chainID, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot update delegators reward map", err)
+	}
+
+	return nil
+}
+
+// calculateProviderDelegatorsRewards returns the provider reward amount and updates the delegator reward map
+func (k Keeper) calculateProviderDelegatorsRewards(ctx sdk.Context, provider sdk.AccAddress, chainID string, totalReward math.Int) (math.Int, error) {
+	stakeEntry, found, _ := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, chainID, provider)
+	if !found {
+		return math.Int{}, utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	// sanity check
+	if stakeEntry.DelegateCommission > 100 {
```

Comment:
this is checked when we stake, should we really check it here?
if we do, lets use 100 and and print an error

Reply:
also feels like this method is redundent, only fetches the stakeentry and calls another function.
lets merge them?

Reply:
done. see https://github.com/lavanet/lava/commit/aaed62b60677647104a3761f38c86c333d59a3a3

Reply:
I think this check is unnecessary (we do not verify all values/params all the time, do we?). 
But if the check is to remain, that it belongs where the value is used, i.e. `CalcProviderReward()`.
Thus, this function can be dropped altogether.

Reply:
dropped the check. see https://github.com/lavanet/lava/commit/c23f6d56950de9bd0d143d7fb62471402a5354e2